### PR TITLE
Verifying signatures

### PR DESCRIPTION
**Pull request by cserb**
- Adding signature to transaction and stake
- Refactoring

### CODE REVIEW

File: spec/transaction_spec.cr [hunk 1]
```diff
@@ -1,20 +1,26 @@
...
   let(:hash) { "c9b0f2181f2783594c8c30f79ef8ff6231494ab50013ed0bb0fc2cd75408f791" }
   let(:from) { "Me" }
   let(:to) { "You" }
   let(:amount) { 100_u64 }
   let(:timestamp) { 1449970561 }
+  let(:signature) { Ledger::Action::Signature.new(
+    v: "034ce3a20d210dc22b79a0944a9b9ef29f3aa50730cc27be7ec4adc601cbcb7372",
+    r: "70189975654209893549254487705918808402318817289942319050850973295297592779325",
+    s: "56643761485509572745767837990785422271240332652515523415468197818355438975767"
+  ) }
```
Comment (q9f) [inline spec/transaction_spec.cr:13]: how did you generate this? `eth_sign`?

File: src/cocol/node/ledger.cr [hunk 2]
```diff
@@ -42,14 +48,14 @@ module Ledger
           **timespan_from_height(height: height),
           wanted_timespan: RETARGET_TIMESPAN,
           current_target: CCL::Pow::Utils.calculate_target(
-            from: Ledger::Repo.blocks[tip_hash].as(Model::Block::Pow).nbits
+            from: Ledger::Repo.blocks[tip_hash].as(Block::Pow).nbits
```
Comment (q9f) [inline src/cocol/node/ledger.cr:51]: what's `nbits`?

File: src/cocol/node/ledger/action.cr [hunk 3]
```diff
@@ -1,34 +1,45 @@
...
 
-  abstract struct AbstractTransaction
+  struct Signature
     include JSON::Serializable
 
+    getter v : String
+    getter r : String
+    getter s : String
+
+    def initialize(@v, @r, @s)
```
Comment (q9f) [inline src/cocol/node/ledger/action.cr:11]: shall we implement this in secp256k1.cr?

https://github.com/q9f/secp256k1.cr/issues/35

File: src/cocol/node/ledger/util.cr [hunk 4]
```diff
@@ -37,6 +39,15 @@ module Ledger
         result: 0_i64)
     end
 
+    def signature_valid?(txn : Ledger::Action::Transaction) : Bool
+      txn_sig = txn.sig.as(Ledger::Action::Signature)
+      r = BigInt.new(txn_sig.r)
+      s = BigInt.new(txn_sig.s)
+      sig = Secp256k1::ECDSA_Signature.new(r: r, s: s)
+      pubdc = Secp256k1::Util.decode_compressed_public_key(txn_sig.v)
```
Comment (q9f) [inline src/cocol/node/ledger/util.cr:47]: ahhhh.... does this comply with `eth_sign`?

File: src/deps.cr [hunk 5]
```diff
@@ -12,6 +12,7 @@ require "http/client"
 require "ccl-pow"
 require "probfin"
 require "ccl-pos"
+require "secp256k1"
```
Comment (q9f) [inline src/deps.cr:15]: :blush:

File: spec/transaction_spec.cr (hunk 1 above)
Comment (cserb) [reply spec/transaction_spec.cr:13]: Manually. I used `crystal play` and required `secp256k1`

File: src/cocol/node/ledger.cr (hunk 2 above)
Comment (cserb) [reply src/cocol/node/ledger.cr:51]: https://bitcoin.org/en/glossary/nbits :)

File: src/cocol/node/ledger/action.cr (hunk 3 above)
Comment (cserb) [reply src/cocol/node/ledger/action.cr:11]: definitely! :)

**Pull request merged**